File: Java/Sorting/Bidirectional_Selection_Sort.java
import java.util.Scanner;

public class Main {
    public static void bidirectionalSelectionSort(int[] arr) {
        int left = 0, right = arr.length - 1;

        while (left < right) {
            int minimum = arr[left];
            int maximum = arr[right];
            int minIndex = left;
            int maxIndex = right;

            for (int i = left +1; i <= right; i++) {
                if (arr[i] < minimum) {
                    minimum = arr[i];
                    minIndex = i;
                }
                if (arr[i] > maximum) {
                    maximum = arr[i];
                    maxIndex = i;
                }
            }

            if (minIndex == right) {
                int temp = arr[left];
                arr[left] = arr[minIndex];
                arr[minIndex] = temp;
            } else {
                int temp = arr[left];
                arr[left] = arr[minIndex];
                arr[minIndex] = temp;

                temp = arr[right];
                arr[right] = arr[maxIndex];
                arr[maxIndex] = temp;
            }

            left++;
            right--;
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter the number of elements in the array: ");
        int n = scanner.nextInt();

        int[] array = new int[n];

        System.out.print("Enter the elements: ");
        for (int i = 0; i < n; i++) {
            array[i] = scanner.nextInt();
        }

        System.out.print("Original array: ");
        for (int i = 0; i < n; i++) {
            System.out.print(array[i] + " ");
        }
        System.out.println();

        bidirectionalSelectionSort(array);

        System.out.print("Sorted array: ");
        for (int i = 0; i < n; i++) {
            System.out.print(array[i] + " ");
        }
        System.out.println();
    }
}

File: Python/Sorting/Bidirectional_Selection_Sort.py
def bidirectional_selection_sort(arr):
    left = 0
    right = len(arr) - 1

    while left < right:
        minimum = arr[left]
        maximum = arr[right]
        min_index = left
        max_index = right

        for i in range(left, right + 1):
            if arr[i] < minimum:
                minimum = arr[i]
                min_index = i
            if arr[i] > maximum:
                maximum = arr[i]
                max_index = i

        if min_index == right:
            arr[left], arr[min_index] = arr[min_index], arr[left]
        else:
            arr[left], arr[min_index] = arr[min_index], arr[left]
            arr[right], arr[max_index] = arr[max_index], arr[right]

        left += 1
        right -= 1

n = int(input("Enter the number of elements in the array: "))
array = list(map(int, input("Enter the elements: ").split()))

print("Original array:", array)

bidirectional_selection_sort(array)

print("Sorted array:", array)
